WordPress/Classes/ViewRelated/Post/PostPreviewGenerator.swift

```diff
@@ -0,0 +1,99 @@
+import Foundation
+
+@objc
+protocol PostPreviewGeneratorDelegate {
+    func preview(_ generator: PostPreviewGenerator, attemptRequest request: URLRequest)
+    func preview(_ generator: PostPreviewGenerator, loadHTML html: String)
+}
+
+class PostPreviewGenerator: NSObject {
+    let post: AbstractPost
+    var delegate: PostPreviewGeneratorDelegate?
+
+    init(post: AbstractPost) {
+        self.post = post
+        super.init()
+    }
+
+    func generate() {
+        guard let link = post.permaLink,
+            let url = URL(string: link),
+            !post.hasLocalChanges() else {
+                showFakePreview()
+                return
+        }
+
+        guard WordPressAppDelegate.sharedInstance().connectionAvailable else {
+            showFakePreview(message:
+                NSLocalizedString("The internet connection appears to be offline.", comment: "") +
+                    " " +
+                NSLocalizedString("A simple preview is shown below.", comment: "")
+            )
+            return
+        }
+
+        if needsLogin() {
+            attemptAuthenticatedRequest(url: url)
+        } else {
+            attemptUnauthenticatedRequest(url: url)
+        }
+    }
+
+    func previewRequestFailed(error: NSError) {
+        showFakePreview(message:
+            NSLocalizedString("There has been an error while trying to reach your site.", comment: "") +
```

Comment:
I might be nitpicking a bit here, but saying there was an error begs the question: "What was the error?". 
It might be nice to include the error description, or barring that maybe just say the site was unavailable/unreachable.  
Thoughts?